core/ipc/sink/off-heap/src/main/java/org/opennms/core/ipc/sink/offheap/QueueFileOffHeapDispatchQueue.java

```diff
@@ -0,0 +1,537 @@
+                    throw new WriteFailedException(e);
+                }
+
+                size = serializedBatch.length + SERIALIZED_OBJECT_HEADER_SIZE_IN_BYTES;
+                fileCapacityLatch.markFlushNeeded();
+            }
+        }
+
+        if (serializedBatch != null) {
+            try {
+                // This is a critical blocking call and it has to be done outside the context of any shared lock with
+                // dequeue() otherwise it will cause a deadlock
+                //
+                // After unblocking we will pick up the lock again and double check that we still need to flush and then
+                // perform that while holding the lock
+                fileCapacityLatch.waitForCapacity(size);
+
+                synchronized (offHeapMutex) {
+                    if (!fileCapacityLatch.isFlushNeeded()) {
+                        return EnqueueResult.DEFERRED;
+                    }
+
+                    try {
+                        offHeapQueue.add(serializedBatch);
+
+                        // Since we just wrote to disk, we need to check the file again to record the current capacity
+                        checkFileSize();
+                    } catch (IOException e) {
+                        throw new WriteFailedException(e);
+                    }
+                }
+            } catch (InterruptedException e) {
+                throw new WriteFailedException(e);
+            }
+        }
+
+        return EnqueueResult.DEFERRED;
+    }
+
+    /**
+     * On every call to dequeue, if the off-heap queue is configured, we check the file for an entry and drain it to the
+     * in-memory queue provided there is room. We then take exclusively from the head of the in-memory queue which
+     * ensures ordering with respect to the two discrete queues.
+     * <p>
+     * After completely draining the queue on disk we check the existing batch for entries and drain them next.
+     */
+    @Override
+    public Map.Entry<String, T> dequeue() throws InterruptedException {
+        LOG.debug("Dequeueing an entry from queue with current size {}", getSize());
+
+        // If off-heap queueing is enabled we need to first check if there is anything to read off-heap
+        if (offHeapQueue != null) {
+            synchronized (offHeapMutex) {
+                // Try to move a batch from the off-heap queue to the in-memory queue
+                if (offHeapQueue.size() > 0 && inMemoryQueue.remainingCapacity() >= batchSize) {
+                    LOG.trace("Found an entry off-heap and there was room in-memory, moving it");
+
+                    try {
+                        byte[] entry = offHeapQueue.peek();
+                        if (entry != null) {
+                            offHeapQueue.remove();
+
+                            try {
+                                inMemoryQueue.addAll(unbatchSerializedBatch(new SerializedBatch(entry)));
```

Comment:
I'm not sure if I understand this code completely.   If there is only one thread that does `dequeue` the order should be maintained. Would like to see a test case.

Reply:
Good idea, I will add a test case to verify the ordering. Since batches are synchronized the order of contents should be guaranteed to be in the same order.

Let me know if you want me to elaborate on how anything works.

Reply:
Adding a test case with some data on file, some data in memory would be sufficient. Then I can play with that and understand the logic better.

Reply:
Looks like there was an issue with the ordering, will fix. Thanks for catching this.

Reply:
If I understand this correctly,  If `dequeue` is called from two different threads, it would pull two different batches of messages in parallel then order may not be maintained when they dispatch in two different threads.

Reply:
All calls to `dequeue` end up taking from the inMemoryQueue data structure and ordering is ensured there by guarding manipulation of that queue with a mutex. It is not possible for two threads to be simultaneously retrieving two different batches at once.

This means that entries are always dequeued in order.

However it is still possible to dispatch entries in a different order since there is multiple consuming threads calling `dequeue` since all the threads ultimately race to finish dispatching. If we need that to be guaranteed to be in order we need make sure there is only one thread consuming from the queue which will likely drastically impact the performance of the AsyncDispatcher.

Note that that race condition existed in the previous implementation so this PR is not violating any previously established ordering guarantee.

Reply:
> All calls to `dequeue` end up taking from the inMemoryQueue data structure and ordering is ensured there by guarding manipulation of that queue with a mutex. It is not possible for two threads to be simultaneously retrieving two different batches at once.
> 
> This means that entries are always dequeued in order.
> 
> However it is still possible to dispatch entries in a different order since there is multiple consuming threads calling `dequeue` since all the threads ultimately race to finish dispatching. If we need that to be guaranteed to be in order we need make sure there is only one thread consuming from the queue which will likely drastically impact the performance of the AsyncDispatcher.
> 
> Note that that race condition existed in the previous implementation so this PR is not violating any previously established ordering guarantee.

Got it. I think it does maintain order. Can we modify test case for verifying that multiple threads doing `dequeue` also maintains order. Probably synchronize both `dequeue` and putting elements back.

I think we don't need to worry about dispatching in order since even after pushing them to kafka, there is no guarantee in order.

Reply:
Multiple threads calling dequeue won't necessarily be in order since there is a race condition on which thread will get to `take()` from the queue first. This should be acceptable since we don't have an ordering guarantee on the consumption side anyways (dispatching).

The queue itself will dispense items in the correct order but there is no way of knowing the order the threads will actually dequeue in so there really isn't any predictable behaviour to test without synchronizing the entire dequeue & process task.

Reply:
yeah, the dequeue method is not synchronized. That's fine then.